distributed/shuffle/_worker_extension.py

```diff
@@ -192,6 +151,336 @@ def heartbeat(self) -> dict[str, Any]:
+    """State for a single active rechunk execution
+
+    This object is responsible for splitting, sending, receiving and combining
+    data shards.
+
+    It is entirely agnostic to the distributed system and can perform a shuffle
+    with other `Shuffle` instances using `rpc` and `broadcast`.
+
+    The user of this needs to guarantee that only `Shuffle`s of the same unique
+    `ShuffleID` interact.
+
+    Parameters
+    ----------
+    worker_for:
+        A mapping partition_id -> worker_address.
+    output_workers:
+        A set of all participating worker (addresses).
+    old:
+        Existing chunking of the array per dimension.
+    new:
+        Desired chunking of the array per dimension.
+    id:
+        A unique `ShuffleID` this belongs to.
+    run_id:
+        A unique identifier of the specific execution of the shuffle this belongs to.
+    local_address:
+        The local address this Shuffle can be contacted by using `rpc`.
+    directory:
+        The scratch directory to buffer data in.
+    nthreads:
+        How many background threads to use for compute.
+    loop:
+        The event loop.
+    rpc:
+        A callable returning a PooledRPCCall to contact other Shuffle instances.
+        Typically a ConnectionPool.
+    scheduler:
+        A PooledRPCCall to to contact the scheduler.
+    memory_limiter_disk:
+    memory_limiter_comm:
+        A ``ResourceLimiter`` limiting the total amount of memory used in either
+        buffer.
+    """
+
+    def __init__(
+        self,
+        worker_for: dict[NIndex, str],
+        output_workers: set,
+        old: ChunkedAxes,
+        new: ChunkedAxes,
+        id: ShuffleId,
+        run_id: int,
+        local_address: str,
+        directory: str,
+        nthreads: int,
+        rpc: Callable[[str], PooledRPCCall],
+        scheduler: PooledRPCCall,
+        memory_limiter_disk: ResourceLimiter,
+        memory_limiter_comms: ResourceLimiter,
+    ):
+        from dask.array.rechunk import _old_to_new
+
+        super().__init__(
+            id=id,
+            run_id=run_id,
+            output_workers=output_workers,
+            local_address=local_address,
+            directory=directory,
+            nthreads=nthreads,
+            rpc=rpc,
+            scheduler=scheduler,
+            memory_limiter_comms=memory_limiter_comms,
+            memory_limiter_disk=memory_limiter_disk,
+            dump=dump_shards,
+            load=load_partition,
+        )
+        self.old = old
+        self.new = new
+        partitions_of = defaultdict(list)
+        for part, addr in worker_for.items():
+            partitions_of[addr].append(part)
+        self.partitions_of = dict(partitions_of)
+        self.worker_for = worker_for
+        self._slicing = rechunk_slicing(old, new)
+        self._old_to_new = _old_to_new(old, new)
+
+    async def _receive(self, data: list[tuple[tuple[NIndex, NIndex], bytes]]) -> None:
```

Comment:
```suggestion
    async def _receive(self, data: list[tuple[TransferShardIDType, bytes]]) -> None:
```

Does this work?

Reply:
What's the best practice here when having concrete subclasses of generics? Using the concrete value of the `TypeVar` or the `TypeVar` itself?

Reply:
cc @graingert/@crusaderky, are you aware of anything here?

Reply:
Turns out that this is necessary and `mypy` does resolve `TransferShardIDType` to the concrete type.